File: sdks/java/conformance-tests/src/test/java/io/croniq/runner/conformance/BodyMatcher.java
package io.croniq.runner.conformance;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.List;
import java.util.Map;

/**
 * Subset matcher with one wildcard token. Used to assert that a recorded
 * request body (parsed as a {@link JsonNode}) satisfies the YAML
 * {@code body_match} spec.
 *
 * <p>Rules:
 *
 * <ul>
 *   <li>YAML {@code null} → actual must be JSON {@code null}.
 *   <li>YAML {@code "*"} → actual must not be null or empty.
 *   <li>Map (YAML object) → every listed key must exist with a matching value;
 *       extra keys in the actual body are ignored.
 *   <li>List (YAML array) → lengths must match; each element must match.
 *   <li>Scalar (string / number / boolean) → exact equality.
 * </ul>
 *
 * <p>Returns {@code null} on success; otherwise a path-rooted error like
 * {@code "$.runner_id: expected 'exec-001' but got 'exec-002'"}.
 */
final class BodyMatcher {

    private BodyMatcher() {}

    static String match(Object expected, JsonNode actual) {
        return matchAt("$", expected, actual);
    }

    private static String matchAt(String path, Object expected, JsonNode actual) {
        if (expected == null) {
            if (actual == null || actual.isNull()) {
                return null;
            }
            return path + ": expected null but got " + actual.asText();
        }
        if (expected instanceof String s) {
            if ("*".equals(s)) {
                if (actual == null || actual.isNull()) {
                    return path + ": expected non-empty (*) but was null";
                }
                if (actual.isTextual() && actual.asText().isEmpty()) {
                    return path + ": expected non-empty (*) but was empty string";
                }
                return null;
            }
            if (actual == null || !actual.isTextual()) {
                return path + ": expected string '" + s + "' but got " + describe(actual);
            }
            return s.equals(actual.asText()) ? null : path + ": expected '" + s + "' but got '" + actual.asText() + "'";
        }
        if (expected instanceof Boolean b) {
            if (actual == null || !actual.isBoolean()) {
                return path + ": expected bool " + b + " but got " + describe(actual);
            }
            return b == actual.asBoolean() ? null : path + ": expected " + b + " but got " + actual.asBoolean();
        }
        if (expected instanceof Number n) {
            if (actual == null || !actual.isNumber()) {
                return path + ": expected number " + n + " but got " + describe(actual);
            }
            // Compare as long when possible, else double — the YAML loader
            // doesn't preserve the original literal type so widen with care.
            if (n.longValue() == actual.asLong() && n.doubleValue() == actual.asDouble()) {
                return null;
            }
            return path + ": expected " + n + " but got " + actual.asText();
        }
        if (expected instanceof Map<?, ?> m) {
            if (actual == null || !(actual instanceof ObjectNode)) {
                return path + ": expected object but got " + describe(actual);
            }
            for (var e : m.entrySet()) {
                String key = e.getKey().toString();
                JsonNode child = actual.get(key);
                String err = matchAt(path + "." + key, e.getValue(), child);
                if (err != null) {
                    return err;
                }
            }
            return null;
        }
        if (expected instanceof List<?> list) {
            if (actual == null || !(actual instanceof ArrayNode arr)) {
                return path + ": expected array but got " + describe(actual);
            }
            if (arr.size() != list.size()) {
                return path + ": expected length " + list.size() + " but got " + arr.size();
            }
            for (int i = 0; i < list.size(); i++) {
                String err = matchAt(path + "[" + i + "]", list.get(i), arr.get(i));
                if (err != null) {
                    return err;
                }
            }
            return null;
        }
        return path + ": unsupported expected type " + expected.getClass().getName();
    }

    private static String describe(JsonNode n) {
        if (n == null) {
            return "missing";
        }
        if (n.isNull()) {
            return "null";
        }
        return n.getNodeType() + " '" + n.asText() + "'";
    }
}

File: sdks/go/conformance/body_matcher.go
package conformance

import (
	"encoding/json"
	"fmt"
)

// MatchBody asserts that `actual` (raw JSON bytes) satisfies the subset
// rules in `expected` (parsed from YAML — map[string]any / []any /
// scalars). Returns nil on success, or a human-readable path-rooted diff
// for the first mismatch found.
//
// One wildcard token is supported: a literal string `"*"` matches any
// non-empty value of any JSON kind. Subset semantics — keys in the
// actual body that aren't mentioned in `expected` are ignored. Array
// length must match.
func MatchBody(expected any, actualJSON string) string {
	if expected == nil {
		return ""
	}
	var actual any
	if actualJSON == "" {
		return "$: empty actual body"
	}
	if err := json.Unmarshal([]byte(actualJSON), &actual); err != nil {
		return fmt.Sprintf("$: not valid JSON: %v", err)
	}
	return match(expected, actual, "$")
}

func match(expected, actual any, path string) string {
	if expected == nil {
		if actual == nil {
			return ""
		}
		return fmt.Sprintf("%s: expected null but got %T", path, actual)
	}

	if s, ok := expected.(string); ok && s == "*" {
		// non-empty wildcard
		switch v := actual.(type) {
		case nil:
			return fmt.Sprintf("%s: expected non-empty wildcard match but got null", path)
		case string:
			if v == "" {
				return fmt.Sprintf("%s: expected non-empty string but got empty", path)
			}
		}
		return ""
	}

	switch e := expected.(type) {
	case map[string]any:
		am, ok := actual.(map[string]any)
		if !ok {
			return fmt.Sprintf("%s: expected object but got %T", path, actual)
		}
		for k, v := range e {
			child, present := am[k]
			if !present {
				return fmt.Sprintf("%s.%s: missing key", path, k)
			}
			if msg := match(v, child, path+"."+k); msg != "" {
				return msg
			}
		}
		return ""

	case []any:
		al, ok := actual.([]any)
		if !ok {
			return fmt.Sprintf("%s: expected array but got %T", path, actual)
		}
		if len(al) != len(e) {
			return fmt.Sprintf("%s: expected %d item(s) but got %d", path, len(e), len(al))
		}
		for i := range e {
			if msg := match(e[i], al[i], fmt.Sprintf("%s[%d]", path, i)); msg != "" {
				return msg
			}
		}
		return ""

	case string:
		s, ok := actual.(string)
		if !ok {
			return fmt.Sprintf("%s: expected string %q but got %T", path, e, actual)
		}
		if s != e {
			return fmt.Sprintf("%s: expected %q but got %q", path, e, s)
		}
		return ""

	case bool:
		b, ok := actual.(bool)
		if !ok {
			return fmt.Sprintf("%s: expected bool but got %T", path, actual)
		}
		if b != e {
			return fmt.Sprintf("%s: expected %v but got %v", path, e, b)
		}
		return ""

	case int:
		return matchNumber(float64(e), actual, path)
	case int64:
		return matchNumber(float64(e), actual, path)
	case float64:
		return matchNumber(e, actual, path)

	default:
		return fmt.Sprintf("%s: unsupported expected type %T", path, expected)
	}
}

func matchNumber(expected float64, actual any, path string) string {
	switch n := actual.(type) {
	case float64:
		if n != expected {
			return fmt.Sprintf("%s: expected %v but got %v", path, expected, n)
		}
		return ""
	case int:
		if float64(n) != expected {
			return fmt.Sprintf("%s: expected %v but got %v", path, expected, n)
		}
		return ""
	default:
		return fmt.Sprintf("%s: expected number but got %T", path, actual)
	}
}
